tools/run_tests/xds_k8s_test_driver/tests/url_map/affinity_test.py

```diff
@@ -102,11 +102,116 @@ def rpc_distribution_validate(self, test_client: XdsTestClient):
+        rpc_distribution = self.configure_and_send(
+            test_client,
+            rpc_types=[RpcTypeEmptyCall, RpcTypeUnaryCall],
+            num_rpcs=_NUM_RPCS)
+        self.assertEqual(3, rpc_distribution.num_peers)
+        self.assertLen(
+            test_client.find_subchannels_with_state(
+                _ChannelzChannelState.READY),
+            3,
+        )
+
+
+@absltest.skipUnless('cpp-client' in xds_k8s_flags.CLIENT_IMAGE.value or \
+                     'java-client' in xds_k8s_flags.CLIENT_IMAGE.value,
+                     'Affinity is currently only implemented in C++ and Java.')
+class TestHeaderBasedAffinityMultipleHeaders(
+        xds_url_map_testcase.XdsUrlMapTestCase):
+
+    @staticmethod
+    def client_init_config(rpc: str, metadata: str):
+        # Config the init RPCs to send with the same set of metadata. Without
+        # this, the init RPCs will not have headers, and will pick random
+        # backends (behavior of RING_HASH). This is necessary to only one
+        # sub-channel is picked and used from the beginning, thus the channel
+        # will only create this one sub-channel.
+        return 'EmptyCall', 'EmptyCall:%s:%s' % (_TEST_METADATA_KEY,
+                                                 _TEST_METADATA_VALUE_EMPTY)
+
+    @staticmethod
+    def url_map_change(
+            host_rule: HostRule,
+            path_matcher: PathMatcher) -> Tuple[HostRule, PathMatcher]:
+        # Update default service to the affinity service.
+        path_matcher["defaultService"] = GcpResourceManager(
+        ).affinity_backend_service()
+        return host_rule, path_matcher
+
+    def xds_config_validate(self, xds_config: DumpedXdsConfig):
+        # 3 endpoints in the affinity backend service.
+        self.assertNumEndpoints(xds_config, 3)
+        self.assertEqual(
+            xds_config.rds['virtualHosts'][0]['routes'][0]['route']
+            ['hashPolicy'][0]['header']['headerName'], _TEST_METADATA_KEY)
+        self.assertEqual(xds_config.cds[0]['lbPolicy'], 'RING_HASH')
+
+    def rpc_distribution_validate(self, test_client: XdsTestClient):
+        rpc_distribution = self.configure_and_send(test_client,
+                                                   rpc_types=[RpcTypeEmptyCall],
+                                                   metadata=_TEST_METADATA,
+                                                   num_rpcs=_NUM_RPCS)
+        # Only one backend should receive traffic, even though there are 3
+        # backends.
+        self.assertEqual(1, rpc_distribution.num_peers)
+        self.assertLen(
+            test_client.find_subchannels_with_state(
+                _ChannelzChannelState.READY),
+            1,
+        )
+        self.assertLen(
+            test_client.find_subchannels_with_state(_ChannelzChannelState.IDLE),
+            2,
+        )
+        emptyCallPeer = list(rpc_distribution.raw['rpcsByMethod']['EmptyCall']
```

Comment:
nit: Python prefers snake case for local variables.

Reply:
Done